### PR TITLE
Refactor to adopt operators and computation graph

### PR DESCRIPTION
**Pull request by KevinXuxuxu**
Refactor Existing components into operators to enable more flexible network structure and computation graph. Did some intense OOD but the basic idea is as follows:
- Base `Operator` class supports:
    - input/output shape: `ishape` and `oshape`
    - predecessors and successors for connecting operators to form graph: `preds` and `succs`
    - Pred and succ counter for topological order forward pass and back prop.
- `ParameterizedOperator` for operators with parameters to be trained (`MatMul`, `Bias`), which support an extra learning rate attribute `rate` which can be tuned per operator or globally.
- `BroadcastOperator` for operators with following property (basically all common activations functions):
    - Doesn't care about input shape and will output the same shape as input
    - Basically does something to a single element, that will be broadcasted to the whole input
- `Normalizer` for normalizing operators like `Softmax` and `BN` (not supported in this PR)
- `Loss` for loss functions (`MSE`, `CrossEntropy`, `CrossEntropyWithSoftmax`):
    - Only allowed as the end of graph (no successors)
    - Only 2 inputs are allowed: model output and ground truth

Please refer to `operator_test` for some example graph constructed with operators. I know this is a huge change, so please feel free to leave any comments.

Followups:
- Better approach for getting grad for Softmax?
- Add `Input` and `GroundTruth` operator
- Add training support and integrate with or refactor model/layer
- Benchmark solutions before/after refactor

### CODE REVIEW

File: src/operator/operator.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            self._forward_pass()
+            self.forward_pass_count = 0
+
+    def back_prop(self):
+        self.back_prod_count += 1
+        if self.back_prod_count == len(self.succs):
+            self._back_prop()
+            self.back_prod_count = 0
+
+    def register(self, pred: 'Operator'):
```
Comment (haozech) [inline src/operator/operator.py]: 感觉可以`register`可以换个更直观的名字，比如`set_input`或者`set_pred`

File: test/operator_test.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        a[7].register(a[6])
+        a[8].register(a[7])
+        a[0].output = 0
+        for succ in a[0].succs:
+            succ.forward_pass()
+        for i, x in enumerate(a):
+            self.assertEqual(x.output, expected_result[i])
+            self.assertEqual(x.forward_pass_count, 0)
+
+    def _prepare_test_case(self) -> List[Operator]:
```
Comment (haozech) [inline test/operator_test.py:49]: 我理解我们的组网方式是以控制operator本身来做的，而不是现在比较主流的控制tensor数据？这个我想了想感觉也能work，但不确定后面是不是会踩些坑

File: test/operator_test.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        # _input - MatMul1 - Bias1 - Sigmoid1 - MatMul2 - Bias2 - Sigmoid2 - MSE
+        #                                                                  /
+        #                                                    _ground_truth
+        ParameterizedOperator.rate = 1.
+        _input = Operator([], [2])
+        _input.output = np.array([[.7, 2.]])
+        _ground_truth = Operator([], [2])
+        _ground_truth.output = np.array([[1., 0.]])
+        mm1 = MatMul([2], [3])
```
Comment (haozech) [inline test/operator_test.py]: 这一段组网的代码有点不直观，能否优化到类似
`a1 = Sigmoid(b1)` 这样的语句

File: src/operator/operator.py (hunk 1 above)
Comment (haozech) [reply src/operator/operator.py]: 另外需要考虑有多输出的pred_operator,确定以它的第几个输出数据作为此operator的输入。例如`split` operator

File: src/operator/common.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
+
+from typing import List
+from src.operator.operator import Operator, ParameterizedOperator
+from src.utils import check_shape
+
+
+class MatMul(ParameterizedOperator):
```
Comment (luyueheng) [inline src/operator/common.py:8]: Suggest changing class name to `Linear`

Comment (luyueheng) [inline src/operator/common.py]: Suggest changing class name to `Linear`

File: src/operator/loss.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    def register_ground_truth(self, pred: Operator):
+        if self.preds[1] is not None:
+            raise Exception('Ground truth for loss is already set')
+        check_shape(self.ishape, pred.oshape)
+        self.preds[1] = pred
+        pred.succs.append(self)
+
+    @property
+    def y(self):
```
Comment (luyueheng) [inline src/operator/loss.py]: Suggest changing `y` to `y_pred` or `y_hat`, to make it explicit as the predicted output y

File: src/operator/operator.py (hunk 1 above)
Comment (KevinXuxuxu) [reply src/operator/operator.py]: - 名字我会改一下
- 多输出和多输入的情况我考虑的是在operator子类里对`register`等方法作重载，加入一些verify shape的逻辑，总体上来讲和其他的operator类似。至于对应的问题我暂时的想法是从简，就是比如对于`split`来说，`succs`里面的operator会按顺序拿到output里面的数据。不过具体做法还可以讨论。

File: test/operator_test.py (hunk 3 above)
Comment (KevinXuxuxu) [reply test/operator_test.py]: 这样确实会更直观，但是考虑到operator的constructor可能还要别的参数（shape, learning_rate, etc.），而且python好像并不支持根据function signature的多态（e.g. Java)，可能会比较难以实现，或者实现出来会很丑....

Comment (KevinXuxuxu) [reply test/operator_test.py]: 硬要写的话只能这样：https://www.geeksforgeeks.org/what-is-a-clean-pythonic-way-to-have-multiple-constructors-in-python/

File: src/operator/common.py (hunk 4 above)
Comment (KevinXuxuxu) [reply src/operator/common.py]: Done

File: src/operator/loss.py (hunk 5 above)
Comment (KevinXuxuxu) [reply src/operator/loss.py]: Changed to `y_pred`

**Review by luyueheng — APPROVED** [review]

**Review by haozech — APPROVED** [review]
LGTM
